### ISSUE TITLE
Named parameter pattern matching

### ISSUE THREAD

**Issue by snoble** [johnynek/bosatsu#240]
I think you've mentioned this before but I couldn't find an issue. I think there's enough pattern matching that this would eliminate the need for accessors. And it's pretty easy to translate these into position based pattern matches

Syntax could be:
```
struct File(name: String, modified_time: Int, size: Int, hash: HashResult)

File(hsh@hash) = file

def time_plus_size(File(mt@modified_time, sze@size)):
  mt + sze
```
Alternatives to @ could be = or <- or ~

**Comment by snoble** [comment]
One neat thing would be if an Enum has the same named param for all structs you could match directly on it

**Comment by johnynek** [comment]
I want to do this, but I think of matching as the dual of assignment or let bindings so I am very partial to the = syntax. 
```
struct Foo(bar: Int)

x = Foo(bar = 3)
Foo(bar = b) = x
```
I think would be the most regular. What do you think?

**Comment by snoble** [comment]
This makes so much sense from a consistency basis. But the gears in my brain are grinding on seeing `=` do right hand assignment. Logically I get why it's correct though

**Comment by snoble** [comment]
Does anything break if when a constructor is being used for deconstruction the label flips sides?
So
`Foo(b = bar) = x`
And with an extra layer
```
bz = Baz(foo = x)
Baz(Foo(b = bar) = foo) = bz
```

It means there are some places where people could have copied and pasted where they can't anymore.

**Comment by johnynek** [comment]
That's interesting: reversing the polarity of the `=` when we are destructuring....

On the other hand, the motivation or `Foo(bar = b) = x` is clearer with: `Foo(bar = b) = Foo(bar = 3)` in this picture, The pattern matching, to me, looks right: `b = 3`.

I'm on the fence with this one. Your suggestion has the strong merit that the thing being bound is always on the left of the `=` which is nice.

**Comment by snoble** [comment]
Maybe you could poll some Galiano retreaters

**Comment by johnynek** [comment]
@snoble I had a chat with @nelhage and @avibryant 

1. they both agreed `Foo(bar = b) = f` is confusing that `b` is the thing that is getting a binding.
2. Nelson noted that javascript actually has object destructuring that works like: `var {bar: b} = {bar: 3}` to bind `b`, but the feeling that `:` was used was less confusing.
3. I felt that since both `=` and `:` don't have any natural directionality, it is hard to parse. We could use `foo = Foo(bar <- 3)` and `Foo(b <- bar) = foo` to add directionality. Candidates might be `<-` and `:=` which are directional.
4. even the directionality doesn't totally sit well with me, because javascript seems to buy the argument that things match so, they would say: `foo = Foo(bar := 3)` and `Foo(bar := b) = foo` or `Foo(b =: bar) = foo` but we should still be pointing at the field it seems to me.

What do you think @snoble?

**Comment by snoble** [comment]
I'm definitely in a place when I think about this that I can see the argument for all these options, so in a sense there isn't a really bad choice from my perspective.

One thing I did notice when reading over the last few comments is that when there's multiple `=`s on a line I have to do a little more work to figure out what's part of the pattern and what's the assignment of the statement. But when there's a `:=` (or any other symbol) this little bit of mental tax goes away.

Your 4th point feels important, and flipping is feeling wrong the more I see it (even though I raised it as an option).

Fwiw, I think if `=` were to be used everywhere it would feel normal very quickly in context. I've just become conditioned to seeing it as directional.

Another option is to not have a symbol and always place the field name first and then have a space. So you'd have `foo = Foo(bar 3, car "two")` and `Foo(bar b) = foo`

**Comment by johnynek** [comment]
interesting to forgo a symbol all together.... I think the parsing should be doable. Looks a bit like haskell to me at that point (Haskell function calls). It does look pretty clean. Not so much visual noise. I think that's a pretty nice suggestion.

**Comment by johnynek** [comment]
closing in favor of #289

**Comment by johnynek** [comment]
see https://www.python.org/dev/peps/pep-0622/ for how python is thinking about pattern matching.
